src/function/table/system/attach_duckdb.cpp

```diff
@@ -0,0 +1,241 @@
+class QueryResultIterator {
+public:
+	explicit QueryResultIterator(QueryResult *result_p) : current_row(*this, 0), result(result_p), base_row(0) {
+		if (result) {
+			chunk = shared_ptr<DataChunk>(result->Fetch().release());
+			if (!chunk) {
+				result = nullptr;
+			}
+		}
+	}
+
+	QueryResultRow current_row;
+	shared_ptr<DataChunk> chunk;
+	QueryResult *result;
+	idx_t base_row;
+
+public:
+	void Next() {
+		if (!chunk) {
+			return;
+		}
+		current_row.row++;
+		if (current_row.row >= chunk->size()) {
+			base_row += chunk->size();
+			chunk = result->Fetch();
+			current_row.row = 0;
+			if (!chunk || chunk->size() == 0) {
+				// exhausted all rows
+				base_row = 0;
+				result = nullptr;
+				chunk.reset();
+			}
+		}
+	}
+
+	QueryResultIterator &operator++() {
+		Next();
+		return *this;
+	}
+	bool operator!=(const QueryResultIterator &other) const {
+		return result != other.result || base_row != other.base_row || current_row.row != other.current_row.row;
+	}
+	const QueryResultRow &operator*() const {
+		return current_row;
+	}
+};
+
+Value QueryResultRow::GetValue(idx_t col_idx) const {
+	return iterator.chunk->GetValue(col_idx, row);
+}
+} // namespace __copied_from_query_result_hpp
+using QueryResultIterator = __copied_from_query_result_hpp::QueryResultIterator;
+
+struct ExternalQueryFunctionData : public TableFunctionData {
+	ExternalQueryFunctionData() {
+	}
+
+	bool finished = false;
+	string file_name = "";
+	string table = "";
+	unique_ptr<DuckDB> database = nullptr;
+	unique_ptr<Connection> connection = nullptr;
+	unique_ptr<QueryResult> tablerelation = nullptr;
+	unique_ptr<QueryResultIterator> rowiterator = nullptr;
+};
+
+static unique_ptr<FunctionData> ExternalQueryBind(ClientContext &context, TableFunctionBindInput &input,
+                                                  vector<LogicalType> &return_types, vector<string> &names) {
+
+	auto result = make_unique<ExternalQueryFunctionData>();
+	result->file_name = input.inputs[0].GetValue<string>();
+	result->table = input.inputs[1].GetValue<string>();
+
+	result->database = make_unique<DuckDB>(result->file_name);
+	result->connection = make_unique<Connection>(*result->database);
+
+	result->tablerelation = result->connection->Table(result->table)->Execute();
+
+	for (idx_t i = 0; i < result->tablerelation->names.size(); i++) {
+		return_types.push_back(result->tablerelation->types[i]);
+		names.emplace_back(result->tablerelation->names[i]);
+	}
+	return move(result);
+}
+
+static void ExternalQueryFunction(ClientContext &context, TableFunctionInput &data_p, DataChunk &output) {
+	auto &data = (ExternalQueryFunctionData &)*data_p.bind_data;
+	if (data.finished) {
+		return;
+	}
+
+	if (data.rowiterator == nullptr) {
+		data.rowiterator = make_unique<QueryResultIterator>(data.tablerelation.get());
+	}
+
+	auto dconn = Connection(context.db->GetDatabase(context));
+
+	idx_t count = 0;
+
+	while ((data.rowiterator)->result != nullptr && count < STANDARD_VECTOR_SIZE) {
```

Comment:
Iterating over a result set one-row-at-a-time is extremely slow. This needs to be changed to iterate over data chunks instead. Ideally also in parallel.